src/hidet/runtime/cuda/cudnn.cpp

```diff
@@ -0,0 +1,616 @@
+                CUDNN_TYPE_BACKEND_DESCRIPTOR, 1, &op_graph));
+    // TODO: Is it okay to hardcode the engine to be CUDNN_ATTR_ENGINE_GLOBAL_INDEX 0?
+    // As mentioned here: https://docs.nvidia.com/deeplearning/cudnn/developer/graph-api.html,
+    // Engine selection should be determined based on some heuristics.
+    int64_t gidx = 0;
+    CHECK_CUDNN(cudnnBackendSetAttribute(engine, CUDNN_ATTR_ENGINE_GLOBAL_INDEX,
+                CUDNN_TYPE_INT64, 1, &gidx));
+    CHECK_CUDNN(cudnnBackendFinalize(engine));
+
+    cudnnBackendDescriptor_t engcfg;
+    CHECK_CUDNN(cudnnBackendCreateDescriptor(CUDNN_BACKEND_ENGINECFG_DESCRIPTOR, &engcfg));
+    CHECK_CUDNN(cudnnBackendSetAttribute(engcfg, CUDNN_ATTR_ENGINECFG_ENGINE,
+                CUDNN_TYPE_BACKEND_DESCRIPTOR, 1, &engine));
+    CHECK_CUDNN(cudnnBackendFinalize(engcfg));
+
+    // Set up the execution plan
+    cudnnBackendDescriptor_t plan;
+    CHECK_CUDNN(cudnnBackendCreateDescriptor(CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR, &plan));
+    CHECK_CUDNN(cudnnBackendSetAttribute(plan, CUDNN_ATTR_EXECUTION_PLAN_HANDLE, CUDNN_TYPE_HANDLE, 1, &cur_handle));
+    CHECK_CUDNN(cudnnBackendSetAttribute(plan, CUDNN_ATTR_EXECUTION_PLAN_ENGINE_CONFIG,
+                CUDNN_TYPE_BACKEND_DESCRIPTOR, 1, &engcfg));
+    CHECK_CUDNN(cudnnBackendFinalize(plan));
+
+    int64_t workspaceSize;
+    CHECK_CUDNN(cudnnBackendGetAttribute(plan, CUDNN_ATTR_EXECUTION_PLAN_WORKSPACE_SIZE,
+                CUDNN_TYPE_INT64, 1, NULL, &workspaceSize));
+    
+    void *dev_ptrs[3] = {ptr_x, ptr_w, ptr_y}; // device pointers
+    int64_t uids[3] = {'x', 'w', 'y'};
+    void *workspace = hidet_cuda_malloc_async(workspaceSize, cur_stream);
+
+    cudnnBackendDescriptor_t varpack;
+    CHECK_CUDNN(cudnnBackendCreateDescriptor(CUDNN_BACKEND_VARIANT_PACK_DESCRIPTOR, &varpack));
+    CHECK_CUDNN(cudnnBackendSetAttribute(varpack, CUDNN_ATTR_VARIANT_PACK_DATA_POINTERS,
+                CUDNN_TYPE_VOID_PTR, 3, dev_ptrs));
+    CHECK_CUDNN(cudnnBackendSetAttribute(varpack, CUDNN_ATTR_VARIANT_PACK_UNIQUE_IDS,
+                CUDNN_TYPE_INT64, 3, uids));
+    CHECK_CUDNN(cudnnBackendSetAttribute(varpack, CUDNN_ATTR_VARIANT_PACK_WORKSPACE,
+                CUDNN_TYPE_VOID_PTR, 1, &workspace));
+    CHECK_CUDNN(cudnnBackendFinalize(varpack));
+
+    // Execute the plan
+    CHECK_CUDNN(cudnnBackendExecute(cur_handle, plan, varpack));
+
+    // Cleanup
+    hidet_cuda_free_async(workspace, cur_stream);
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(xDesc));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(wDesc));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(yDesc));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(cDesc));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(fprop));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(op_graph));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(engine));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(engcfg));
+    CHECK_CUDNN(cudnnBackendDestroyDescriptor(plan));
```

Comment:
Might be good to benchmark the performance of our implementation vs. PyTorch's conv2d performance. I am not sure whether the overhead of creating/destroying descriptors is large enough to influence the performance.